### PR TITLE
rgw: ignore SIGXFSZ, which apparently can triggered by heavy ops-log …

### PR DESCRIPTION
**Pull request by mattbenjamin**
…traffic

Fixes: https://tracker.ceph.com/issues/64244

That is, it causes radosgw to ignore SIGXFSZ, which in the apparent reproducer should not cause radosgw to terminate.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
interesting, it looks like this signal is coming from something like `ulimit -f`?
> -f     The maximum size of files written by the shell and its children

from https://www.man7.org/linux/man-pages/man2/setrlimit.2.html
> RLIMIT_FSIZE
              This is the maximum size in bytes of files that the
              process may create.  Attempts to extend a file beyond this
              limit result in delivery of a SIGXFSZ signal.  By default,
              this signal terminates a process, but a process can catch
              this signal instead, in which case the relevant system
              call (e.g., [write(2)](https://www.man7.org/linux/man-pages/man2/write.2.html), [truncate(2)](https://www.man7.org/linux/man-pages/man2/truncate.2.html)) fails with the error
              EFBIG.

does that mean the ops log socket becomes completely unusable when it reaches this size?

**Comment by mattbenjamin** [comment]
it would mean that it does exactly what we, iirc, speculated--it becomes flow-controlled--meaning that attempts to append new data will fail with an error

**Comment by mattbenjamin** [comment]
it should not mean that the socket cannot be read-from--so not completely unusable;  just unable to be written to until (some amount of?) buffer space is recycled

**Comment by mattbenjamin** [comment]
so I take that text to be in support of our existing design, I just was not aware that we needed to handle SIGXFSZ in order to get that behavior

**Comment by pr0ton11** [comment]
Any updates if this could make it to a bugfix release for 18.2? Also does this need to be backported to 17 and 16?

**Comment by cbodley** [comment]
> @mattbenjamin requested a review from @mdw-at-linuxbox 2 months ago

bump!

**Review by dang — APPROVED** [review]

**Comment by cbodley** [comment]
@mattbenjamin do you think this can make squid?

**Comment by mattbenjamin** [comment]
@cbodley it should work correctly on squid--I can create a backport?

**Comment by cbodley** [comment]
> @cbodley it should work correctly on squid--I can create a backport?

@mattbenjamin you have this tagged DNM, so i assumed you were waiting on Marcus' review. should we change that tag to needs-qa?

**Comment by cbodley** [comment]
jenkins test make check

File: src/rgw/rgw_main.cc [hunk 1]
```diff
@@ -134,6 +134,7 @@ int main(int argc, char *argv[])
   register_async_signal_handler(SIGTERM, rgw::signal::handle_sigterm);
   register_async_signal_handler(SIGINT, rgw::signal::handle_sigterm);
   register_async_signal_handler(SIGUSR1, rgw::signal::handle_sigterm);
+  register_async_signal_handler(SIGXFSZ, rgw::signal::sig_handler_noop);
```
Comment (cbodley) [inline src/rgw/rgw_main.cc:137]: https://pulpito.ceph.com/cbodley-2024-04-05_16:34:41-rgw-wip-cbodley-testing-distro-default-smithi/ has some valgrind failures due to a leak here. are we missing a cleanup step?
```xml
<error>
  <unique>0x99108</unique>
  <tid>1</tid>
  <kind>Leak_DefinitelyLost</kind>
  <xwhat>
    <text>144 bytes in 1 blocks are definitely lost in loss record 24 of 40</text>
    <leakedbytes>144</leakedbytes>
    <leakedblocks>1</leakedblocks>
  </xwhat>
  <stack>
    <frame>
      <ip>0x48462F3</ip>
      <obj>/usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>operator new[](unsigned long)</fn>
      <dir>/builddir/build/BUILD/valgrind-3.21.0/coregrind/m_replacemalloc</dir>
      <file>vg_replace_malloc.c</file>
      <line>714</line>
    </frame>
    <frame>
      <ip>0xCE5558</ip>
      <obj>/usr/bin/radosgw</obj>
      <fn>SignalHandler::register_handler(int, void (*)(int), bool)</fn>
    </frame>
    <frame>
      <ip>0x499D81</ip>
      <obj>/usr/bin/radosgw</obj>
      <fn>main</fn>
    </frame>
  </stack>
</error>
```

Comment (mattbenjamin) [reply src/rgw/rgw_main.cc:137]: yes, it looks like there's an unregister_handler, will update

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-04-09_13:16:23-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
> The following tests FAILED:
	 22 - run-tox-cephadm (Failed)

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test windows

**Comment by cbodley** [comment]
> Slave disconnected
The slave crashed or was restarted

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test windows

**Comment by cbodley** [comment]
> No problems were identified. If you know why this problem occurred, please add a suitable Cause for it.

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test windows

**Comment by cbodley** [comment]
jenkins pls :(

**Pull request merged**